source/common/router/router.cc

```diff
@@ -380,8 +396,9 @@ Http::FilterHeadersStatus Filter::decodeHeaders(Http::HeaderMap& headers, bool e
 
   ENVOY_STREAM_LOG(debug, "router decoding headers:\n{}", *callbacks_, headers);
 
-  upstream_request_ = std::make_unique<UpstreamRequest>(*this, *conn_pool);
-  upstream_request_->encodeHeaders(end_stream);
+  UpstreamRequestPtr upstream_request = std::make_unique<UpstreamRequest>(*this, *conn_pool);
+  upstream_request->moveIntoList(std::move(upstream_request), upstream_requests_);
```

Comment:
Seems like we're now keeping track of all the upstream requests even in the non-hedging case, we probably want to remove them from the list when we're done with them to ensure that we're not changing the current, non-hedging behavior.

Reply:
Totally optional and quite possibly overkill, but it might be worth making some of the single->list changes in their own PR where there's no functional change, to make it more clear what the deltas are.

Reply:
I gave this suggestion a try and it seems like a good standalone PR: https://github.com/envoyproxy/envoy/pull/6540/files